File: Python/Linked_List/138. Copy List with Random Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
from typing import Optional
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        dummy = Node(0)
        dummy.next = head

        while head:
            new_node = Node(head.val)
            new_node.next = head.next
            head.next = new_node
            head = new_node.next

        head = dummy.next
        while head:
            if head.random:
                head.next.random = head.random.next
            head = head.next.next
        
        head = dummy.next.next
        dummy.next = head
        while head:
            if head.next:
                head.next = head.next.next
            head = head.next
        
        return dummy.next

File: CPP/LinkedList/138. Copy List with Random Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) return nullptr;

        Node* dummy = new Node(0);
        dummy->next = head;

        while (head != nullptr){
            Node* new_node = new Node(head->val);
            new_node->next = head->next;
            head->next = new_node;
            head = new_node->next;
        }

        head = dummy->next;
        while (head != nullptr){
            if (head->random != nullptr){
                head->next->random = head->random->next;
            }
            head = head->next->next;
        }
        
        Node* origin_head = dummy->next;
        head = dummy->next->next;
        dummy->next = head;
        while (head != nullptr) {
            origin_head->next = head->next;
            if (head->next != nullptr){
                head->next = head->next->next;
            }
            head = head->next;
            origin_head = origin_head->next;
        }

        return dummy->next;
    }
};